docs: add `CONTRIBUTING.md` (#396)
invertase/melos: CONTRIBUTING.md

# Contribution Guidelines

**Note:** If these contribution guidelines are not followed your issue or PR might be closed, so
please read these instructions carefully.

_See also: [Flutter's code of conduct](https://flutter.dev/design-principles/#code-of-conduct)_

## About

Melos is a tool that optimizes the workflow around managing multi-package repositories with git and Pub.

## Contribution types

### Bug Report

- If you find a bug, please first report it using [Github issues](https://github.com/invertase/melos/issues/new?assignees=&labels=bug%2Ctriage&template=bug_report.yml&title=fix%3A++).
  - First check if there is not already an issue for it; duplicated issues will be closed.

### Bug Fix

- If you'd like to submit a fix for a bug, please read the [How To](#how-to-contribute) for how to send a Pull Request.
- Indicate on the open issue that you are working on fixing the bug and the issue will be assigned to you.
- Write `Fix #xxxx` in your PR text, where xxxx is the issue number (if there is one).
- Include a test that isolates the bug and verifies that it was fixed.

### New Features

> - If you'd like to add a feature to the library that doesn't already exist, feel free to describe the feature in a new [GitHub issue](https://github.com/invertase/melos/issues/new?assignees=&labels=bug%2Ctriage&template=bug_report.yml&title=fix%3A++).

Reviewer: ```suggestion
- If you'd like to add a feature to the library that doesn't already exist, feel free to describe the feature in a new [GitHub issue](https://github.com/invertase/melos/issues/new?assignees=&labels=feature+request%2Ctriage&template=feature_request.yml&title=feature%3A++).
```

- If you'd like to implement the new feature, please wait for feedback from the project maintainers before spending too much time writing the code. In some cases, enhancements may not align well with the project future development direction.
- If applicable, implement the code for the new feature and please read the [How To](#how-to-contribute).

### Documentation & Miscellaneous

- If you have suggestions for improvements to the documentation or examples (or something else), we would love to hear about it.

> - As always first file a [Github issue].

Reviewer 2: ```suggestion
- As always first file a [GitHub issue](https://github.com/invertase/melos/issues/new).
```

@blaugold There is no issue template for this kind of issue, right?

Reviewer: That's correct. If you want to add one... :)

> ## Contribution Guidelines

Reviewer: This should be a level 1 heading.

**Note:** If these contribution guidelines are not followed your issue or PR might be closed, so
please read these instructions carefully.

> _See also: [Flutter's code of conduct](https://flutter.dev/design-principles/#code-of-conduct)_

Reviewer: This is Flutter's code of conduct. Not sure if invertase has their own or if we should include Flutter's code of conduct.

cc @Salakar

## About

Melos is a tool that optimizes the workflow around managing multi-package repositories with git and Pub.

## Contribution types

### Bug Report

- If you find a bug, please first report it using [Github issues].

>   - First check if there is not already an issue for it; duplicated issues will be closed.

Reviewer: Is this item meant to be indented?

Author: Yes, I thought this was a part of the above point.

### Bug Fix

- If you'd like to submit a fix for a bug, please read the [How To](#how-to-contribute) for how to send a Pull Request.
- Indicate on the open issue that you are working on fixing the bug and the issue will be assigned to you.
- Write `Fix #xxxx` in your PR text, where xxxx is the issue number (if there is one).
- Include a test that isolates the bug and verifies that it was fixed.

### New Features

> - If you'd like to add a feature to the library that doesn't already exist, feel free to describe the feature in a new [GitHub issue].

Reviewer: There is no URL defined to link `[GitHub issue]`?

> - You can also join us on [Discord]() to discuss some initials thoughts.

Reviewer: We don't have a discord :)

### Documentation & Miscellaneous

- If you have suggestions for improvements to the documentation or examples (or something else), we would love to hear about it.
- As always first file a [Github issue].
- Implement the changes to the documentation, please read the [How To](#how-to-contribute).

## How To Contribute

### Requirements

- Linux, Mac OS X, or Windows.
- [Git](https://git-scm.com) (used for source version control).
- An IDE such as [Android Studio](https://developer.android.com/studio) or [Visual Studio Code](https://code.visualstudio.com/).

### Forking & cloning the repository

- Ensure all the dependencies described in the previous section are installed.
- Fork `https://github.com/invertase/melos` into your own GitHub account. If
  you already have a fork, and are now installing a development environment on
  a new machine, make sure you've updated your fork so that you don't use stale
  configuration options from long ago.
- `git clone git@github.com:<your_name_here>/melos.git`
- `git remote add upstream git@github.com:invertase/melos.git` (So that you
  fetch from the main repository, not your clone, when running `git fetch` or `git pull`

> ## Environment Setup

Reviewer: This heading is not necessary.

### Local development setup

To setup and use this melos mono repo locally for the purposes of contributing, clone it and run the following commands from the root of the repository:

```bash
# Install melos if it's not already installed:
dart pub global activate melos

# Activate 'melos' from path:
melos activate

# Confirm you now using a local development version:
melos --help
# You should now see a banner printed at the top of the help output similar to:
# ---------------------------------------------------------
# | You are running a local development version of melos. |
# ---------------------------------------------------------
```

### Install

```bash
dart pub global activate melos

> ```

Reviewer: This section is not necessary.

Author: This was suggested to be added [here](https://github.com/invertase/melos/issues/394#issuecomment-1270748150)

Reviewer: Yeah, that's a bit confusing. The `Install` section should be above the `Local development setup` section in the REAMDE.

Reviewer: But in `CONTRIBUTING.md` we don't need the `Install` section.

Author: Removed.

Author: > Yeah, that's a bit confusing. The `Install` section should be above the `Local development setup` section in the REAMDE.

Should I create a separate issue for it?

Reviewer: You can just make that change in this PR. Maybe also replace `Local development setup` with `How to contribute` and link to this new document.

Author: Then I shouldn't remove `Local development setup` from `CONTRIBUTING.md`.

> ## Performing changes

Reviewer: This should be a level 3 heading.

- Create a new local branch from `main` (e.g. `git checkout -b my-new-feature`)
- Make your changes (try to split them up with one PR per feature/fix).
- When committing your changes, make sure that each commit message is clear

>  (e.g. `git commit -m 'Added CONTRIBUTING.md'`).

Reviewer: The example commit message should follow the conventional commit spec, e.g. `docs: Added CONTRIBUTING.md`.

- Push your new branch to your own fork into the same remote branch
 (e.g. `git push origin my-username.my-new-feature`, replace `origin` if you use another remote.)

>  ### Open a pull request

Reviewer: Please remove the leading space.

# Contribution Guidelines

**Note:** If these contribution guidelines are not followed your issue or PR might be closed, so
please read these instructions carefully.

_See also: [Flutter's code of conduct](https://flutter.dev/design-principles/#code-of-conduct)_

## About

Melos is a tool that optimizes the workflow around managing multi-package repositories with git and Pub.

## Contribution types

> 

Reviewer 2: ```suggestion
```

Author: Done

### Bug Report

- If you find a bug, please first report it using [Github issues].
  - First check if there is not already an issue for it; duplicated issues will be closed.

> 

Reviewer 2: ```suggestion
```

Author: Done

### Bug Fix

- If you'd like to submit a fix for a bug, please read the [How To](#how-to-contribute) for how to send a Pull Request.
- Indicate on the open issue that you are working on fixing the bug and the issue will be assigned to you.
- Write `Fix #xxxx` in your PR text, where xxxx is the issue number (if there is one).
- Include a test that isolates the bug and verifies that it was fixed.

> 

Reviewer 2: ```suggestion
```

Author: Done

### New Features

- If you'd like to add a feature to the library that doesn't already exist, feel free to describe the feature in a new [GitHub issue].
- You can also join us on [Discord]() to discuss some initials thoughts.
- If you'd like to implement the new feature, please wait for feedback from the project maintainers before spending too much time writing the code. In some cases, enhancements may not align well with the project future development direction.
- If applicable, implement the code for the new feature and please read the [How To](#how-to-contribute).

> 

Reviewer 2: ```suggestion
```

Author: Done

### Documentation & Miscellaneous

- If you have suggestions for improvements to the documentation or examples (or something else), we would love to hear about it.
- As always first file a [Github issue].
- Implement the changes to the documentation, please read the [How To](#how-to-contribute).

## How To Contribute

> 

Reviewer 2: ```suggestion
```

Author: Done

### Requirements

> - Linux, Mac OS X, or Windows.

Reviewer 2: ```suggestion

- Linux, Mac OS X, or Windows.
```

Author: Done

- [Git](https://git-scm.com) (used for source version control).
- An IDE such as [Android Studio](https://developer.android.com/studio) or [Visual Studio Code](https://code.visualstudio.com/).

### Forking & cloning the repository

- Ensure all the dependencies described in the previous section are installed.
- Fork `https://github.com/invertase/melos` into your own GitHub account. If
  you already have a fork, and are now installing a development environment on
  a new machine, make sure you've updated your fork so that you don't use stale
  configuration options from long ago.
- `git clone git@github.com:<your_name_here>/melos.git`
- `git remote add upstream git@github.com:invertase/melos.git` (So that you
  fetch from the main repository, not your clone, when running `git fetch` or `git pull`

> 

Reviewer 2: ```suggestion
```

Author: Done
